### PR TITLE
GRU/AUGRUCell shape inference function

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - Add shape_infer function for GRUCell and AUGRUCell
 - Enablement of GRUCell shape_infer for CPU
 - Enablement of AUGRUCell/Sequence shape_infer for CPU
  ----------------------------------  
   - Shape inference for GRU/AUGRUSequence has been added with PR: 
   https://github.com/openvinotoolkit/openvino/pull/12982
    
### Tickets:
 - 94060

### CODE REVIEW

File: src/core/tests/type_prop/augru_sequence.cpp [hunk 1]
```diff
@@ -281,6 +281,29 @@ TEST(type_prop, augru_sequence_invalid_attention_gate) {
...
+    params.seq_length = 6;
+    params.input_size = 4;
+    params.hidden_size = 128;
+    params.et = element::f32;
+
+    auto augru_sequence = augru_seq_init(params);
+    auto invalid_attention_gate = make_shared<opset9::Parameter>(params.et, PartialShape{999, params.seq_length, 1});
+    augru_sequence->set_argument(6, invalid_attention_gate);
+
+    try {
```
Comment (praasz) [inline src/core/tests/type_prop/augru_sequence.cpp:298]: Use `OV_EXPECT_THROW` here?

File: src/core/shape_inference/include/augru_cell_shape_inference.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    const auto& a_shape = input_shapes[expected_in_shapes_count - 1];
+    const auto& x_shape = input_shapes[0];
+    NODE_VALIDATION_CHECK(op, a_shape.rank().compatible(2), "'A' input must be a 2D tensor.");
+    if (a_shape.rank().is_static()) {
+        if (x_shape.rank().is_static()) {
+            NODE_VALIDATION_CHECK(op,
+                                  x_shape.rank().get_length() > 1 && a_shape[0].compatible(x_shape[0]),
+                                  "Dimension `batch_size` must be the same for `X` and `A` inputs.");
+        }
+        NODE_VALIDATION_CHECK(op, a_shape[1].compatible(1), "The last dimension of `A` shape must be equal to `1`.");
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/augru_cell_shape_inference.hpp:39]: As it is "the last" for the Cell and Sequence, I would like to stay with unified message, and keep the validation of the last dimension at the end.

File: src/common/transformations/src/ngraph_ops/augru_cell.cpp [hunk 3]
```diff
@@ -87,55 +64,20 @@ void ov::op::internal::AUGRUCell::validate_and_infer_types() {
...
-            NODE_VALIDATION_CHECK(this,
-                                  b_pshape[0].compatible(merged_hidden_size * (s_gates_count + m_linear_before_reset)),
-                                  "Parameter hidden_size mistmatched in B input. Current value is: ",
-                                  b_pshape[0].get_length(),
-                                  ", expected: ",
-                                  merged_hidden_size.get_length() * (s_gates_count + m_linear_before_reset),
-                                  ".");
-        }
-    }
+    std::vector<ov::PartialShape> input_shapes = {x_pshape, ht_pshape, w_pshape, r_pshape, b_pshape, a_pshape};
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/src/ngraph_ops/augru_cell.cpp]: Ok, done.

File: src/common/transformations/src/ngraph_ops/augru_cell.cpp [hunk 4]
```diff
@@ -87,55 +64,20 @@ void ov::op::internal::AUGRUCell::validate_and_infer_types() {
...
+    std::vector<ov::PartialShape> output_shapes = {ov::PartialShape{2}};
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/src/ngraph_ops/augru_cell.cpp]: Right, thanks!

File: src/core/shape_inference/include/augru_cell_shape_inference.hpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                          "Incorrect number of input shapes has been provided. Expected: ",
+                          expected_in_shapes_count,
+                          ", got: ",
+                          input_shapes.size(),
+                          ".");
+
+    rnn_seq::gru_cell_shape_infer(op, input_shapes, output_shapes);
+
+    // `A` input shape validation // [batch_size, 1]
+    const auto& a_shape = input_shapes[expected_in_shapes_count - 1];
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/augru_cell_shape_inference.hpp]: Done

File: src/common/transformations/src/ngraph_ops/augru_cell.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/src/ngraph_ops/augru_cell.cpp]: Done

File: src/core/src/op/gru_cell.cpp [hunk 6]
```diff
@@ -87,87 +88,28 @@ bool op::v3::GRUCell::visit_attributes(AttributeVisitor& visitor) {
...
+    const auto& ht_pshape = get_input_partial_shape(1);
+    const auto& w_pshape = get_input_partial_shape(2);
+    const auto& r_pshape = get_input_partial_shape(3);
+    const auto& b_pshape = get_input_partial_shape(4);
 
-    // Mark inputs which are relevant to output parameters
-    set_input_is_relevant_to_shape(0);
-    set_input_is_relevant_to_shape(1);
-    set_input_is_relevant_to_shape(3);
+    std::vector<ov::PartialShape> input_shapes{x_pshape, ht_pshape, w_pshape, r_pshape, b_pshape};
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/op/gru_cell.cpp]: Done

File: src/core/tests/type_prop/augru_sequence.cpp (hunk 1 above)
Comment (mitruska) [reply src/core/tests/type_prop/augru_sequence.cpp]: Sure, done

File: src/core/shape_inference/include/augru_cell_shape_inference.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/shape_inference/include/augru_cell_shape_inference.hpp:39]: But still 2nd dim used instead last, change to:?
```suggestion
        NODE_VALIDATION_CHECK(op, a_shape.back().compatible(1), "The last dimension of `A` shape must be equal to `1`.");
```

[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/augru_cell_shape_inference.hpp:39]: It seems that overload for `back()` is not implemented.
"`PartialShape` has no member named `back`"

**Review by praasz — APPROVED** [review]

**Review by praasz — APPROVED** [review]

**Review by itikhono — APPROVED** [review]

**Pull request merged**